### ISSUE TITLE
Change sub-resource referrer policy to report 3p origin

### ISSUE THREAD

**Issue by pes10k** [brave/brave-browser#8696]
Currently Brave does a number of things to prevent privacy loss through the referrer header

https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)#modified-features-and-functionality

This has significant webcompat cost, which we try to mitigate with an [exception list](https://github.com/brave/referrer-whitelist).

Since Brave made this decision, Chromium has gotten tighter with its handling of the [referrer header](https://www.chromestatus.com/feature/6251880185331712).  Basically, they have a default of `strict-origin-when-cross-origin`; sites who want the previous, more permissive behavior need to opt in.

To reduce the webcompat cost, but still aggressively protect user privacy, we should:

1. Use Chromium's behavior as the default
2. Respect when sites specify a more restrictive policy (e.g. `no-referrer`)
3. Treat less restrictive policies as `strict-origin-when-cross-origin`
4. Never send referrer for insecure requests

E.g.
1. Set effective policy to `strict-origin-when-cross-origin`.
2. Is this a cross-site navigation?
   - `GET` / `HEAD`?
     - If yes, change effective policy to `no-referrer`
   - Otherwise:
     - Requested policy of `origin`?
       - If yes, change effective policy to `strict-origin`
     - Requested policy of `no-referrer`, `same-origin`, or `strict-origin`?
       - If yes, change effective policy to the requested policy. 
3. Is this a sub-resource or iframe request?
   - Requested policy of `origin`?
     - If yes, change effective policy to `strict-origin`
   - Requested policy of `no-referrer`, `same-origin`, or `strict-origin`?
     - If yes, change effective policy to the requested policy.

(previously expressed as)
~* if request is HTTP -> `no-referrer`~
~* elif {`no-referrer-when-downgrade`,`unsafe-url`, `origin`, `origin-when-cross-origin`} -> `strict-origin-when-cross-origin`~
~* elif {`no-referrer`, `same-origin`, `strict-origin`} -> no change~
~* else `strict-origin-when-cross-origin`~

**Interaction with Shields**
This intervention should be tied to the cookie setting toggle in shields.  When shields are down, or the cookies setting for the top level URL is "All cookies allowed", Brave should default to Chromium behavior.  When the cookie setting in shields is "Cross-site cookies blocked" or "All cookies blocked", the policy described in this issue should apply.

**Comment by pes10k** [comment]
Note, re privacy confab discussion, there would be no changes to current policy for navigations.

There is also a medium-to-long term goal of building crawls to build a known-deny list, distinct from this change

**Comment by fmarier** [comment]
How should we deal with a referrer policy of `origin`? If we upgrade to `strict-origin-when-cross-origin`, then it's:
- stronger with respect to HTTP resources
- weaker with respect to same-origin resources

We could instead upgrade to `origin` to `strict-origin`.

**Comment by pes10k** [comment]
I think the diff between origin and strict origin ends up being a NOOP, bc of the first case above (i.e. if request is HTTP -> no-referrer).  But, conceptually, I think `origin` -> `strict-origin` sounds good

**Comment by fmarier** [comment]
https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)#modified-features-and-functionality will need to be updated once this is fixed.

**Comment by pes10k** [comment]
Also, once this is complete, revisit https://github.com/brave/brave-browser/issues/7933 to see if it still needs an independent fix

**Comment by fmarier** [comment]
With the change to top-level `POST` navigations, #5075 and #9470 might get resolved.

**Comment by pes10k** [comment]
Manual tests for QA https://dev-pages.bravesoftware.com/referrer/index.html and https://dev-pages.brave.software/index.html (same tests, but both origins are needed for some of the tests)

**Comment by fmarier** [comment]
I updated my test pages at https://fmarier.github.io/brave-testing/referrer-spoofing.html (including
new form test cases), https://referrer.fmarier.org/redirects.html, and https://referrer.fmarier.org/samesite.html.

I also confirmed that all of the test cases on https://dev-pages.brave.software/referrer/index.html and https://fmarier.com/referrer/strict-origin.html work.

**Comment by fmarier** [comment]
I updated the following wiki pages:
- https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)#modified-features-and-functionality
- https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections

**Comment by LaurenWags** [comment]
Marking as `QA/Blocked` for same reason as https://github.com/brave/brave-browser/issues/6657#issuecomment-670033113

**Comment by pes10k** [comment]
@LaurenWags which issues / tests on https://dev-pages.brave.software/referrer/index.html and https://fmarier.com/referrer/strict-origin.html are not working for you?

**Comment by LaurenWags** [comment]
1. On https://fmarier.github.io/brave-testing/referrer-spoofing.html I don't know that some of the referrer stuff is working as expected:
a. under Form-submission, when I click on the `same-origin submission` button, i get a 405 not allowed message
<img width="1195" alt="Screen Shot 2020-08-06 at 1 07 24 PM" src="https://user-images.githubusercontent.com/28145373/89560939-ccb33280-d7e5-11ea-9ecf-adb7c17df496.png">

b. under iframe navigations the `cross-origin navigation` and `cross-origin navigation (redirected via meta)` initially show as expected. However, if I first click on `cross-origin navigation` link under Top-level navigations, and I navigate back to https://fmarier.github.io/brave-testing/referrer-spoofing.html they do not show expected result until I refresh the page. is this ok?
<img width="665" alt="Screen Shot 2020-08-06 at 1 08 13 PM" src="https://user-images.githubusercontent.com/28145373/89561033-ebb1c480-d7e5-11ea-8f6e-c428d54c0d05.png">


2. On https://referrer.fmarier.org/redirects.html (which I see now isn't in the PR but is listed in https://github.com/brave/brave-browser/issues/8696#issuecomment-644307746) the links at the top do not show blank, I see `Referrer: https://referrer.fmarier.org/`  on each page, expected? (boxes at the bottom of the page all work as expected)
<img width="584" alt="Screen Shot 2020-08-06 at 1 08 53 PM" src="https://user-images.githubusercontent.com/28145373/89561109-084dfc80-d7e6-11ea-8682-ace771a2b8b0.png">

**Comment by fmarier** [comment]
> a. under Form-submission, when I click on the same-origin submission button, i get a 405 not allowed message

Now fixed. This was due to GitHub apparently starting to block POST requests.

> b. under iframe navigations the cross-origin navigation and cross-origin navigation (redirected via meta) initially show as expected. However, if I first click on cross-origin navigation link under Top-level navigations, and I navigate back to https://fmarier.github.io/brave-testing/referrer-spoofing.html they do not show expected result until I refresh the page. is this ok?

Partially fixed (the first test case) by switching away from `whatismyreferrer.com`. I believe we may have an underlying bug here ("cross-origin navigation (redirected via meta)").

> On https://referrer.fmarier.org/redirects.html (which I see now isn't in the PR but is listed in #8696 (comment)) the links at the top do not show blank, I see Referrer: https://referrer.fmarier.org/ on each page, expected? (boxes at the bottom of the page all work as expected)

The links at the top are out-of-scope for this PR. The boxes, which work fine, is all that we should be testing against this PR.

**Comment by LaurenWags** [comment]
Thank you @fmarier! I'll get this all re-checked and log a follow-up issue 😄

**Comment by LaurenWags** [comment]
Verified passed with

```
Brave | 1.12.108 Chromium: 84.0.4147.105 (Official Build) (64-bit)
-- | --
Revision | a6b12dfad6663f13a7e16e9a42a6a4975374096b-refs/branch-heads/4147@{#943}
OS | macOS Version 10.14.6 (Build 18G3020)
```

* Verified test plan from https://github.com/brave/brave-core/pull/5613 
* Confirmed tests on https://fmarier.github.io/brave-testing/referrer-spoofing.html work as expected. Logged https://github.com/brave/brave-browser/issues/11119 as follow up noted in https://github.com/brave/brave-browser/issues/8696#issuecomment-670108015.
* Confirmed tests on https://referrer.fmarier.org/samesite.html work as expected.
* Confirmed tests on https://fmarier.com/referrer/strict-origin.html work as expected.
* Confirmed tests on https://referrer.fmarier.org/redirects.html (bottom boxes only per https://github.com/brave/brave-browser/issues/8696#issuecomment-670108015) work as expected.
* Confirmed tests on https://dev-pages.brave.software/referrer/index.html work as expected.

------------

**Comment by srirambv** [comment]
Verification passed on OnePlus 6T with Android 10 running 1.12.111 x64 build

- Verified test plan from brave/brave-core#5613
- Verified tests on `https://fmarier.github.io/brave-testing/referrer-spoofing.html` and works as expected
- Verified tests on `https://referrer.fmarier.org/samesite.html` and works as expected
- Verified tests on `https://fmarier.com/referrer/strict-origin.html` and works as expected
- Verified tests on `https://referrer.fmarier.org/redirects.html` and works as expected
- Verified tests on `https://dev-pages.brave.software/referrer/index.html` and works as expected
--------
Verification passed on Samsung Tab A with Android 10 running 1.12.111 x64 build

- Verified test plan from brave/brave-core#5613
- Verified tests on `https://fmarier.github.io/brave-testing/referrer-spoofing.html` and works as expected
- Verified tests on `https://referrer.fmarier.org/samesite.html` and works as expected
- Verified tests on `https://fmarier.com/referrer/strict-origin.html` and works as expected
- Verified tests on `https://referrer.fmarier.org/redirects.html` and works as expected
- Verified tests on `https://dev-pages.brave.software/referrer/index.html` and works as expected
